### ISSUE TITLE
Missing synchronization after asynchronous copies?

### ISSUE THREAD

**Issue by L30nardoSV** [ccsb-scripps/AutoDock-GPU#31]
This error was found with [oclgrind](https://github.com/jrprice/Oclgrind/wiki/Using-Oclgrind) and points to the last asynchronous copies in `kernel4` and `kernel_gradient`:

```zsh
Work-item finished without waiting for events.
```

Other kernels might suffer from this too:

* See `master` -> `kernel3` and `kernel4`.
* See `debugfastergrad` -> `kernel3`, `kernel4`, `kernel_gradient`, and `kernel_fire`.

Keep in mind the following information from the [v2.0 standard](https://www.khronos.org/registry/OpenCL/sdk/2.0/docs/man/xhtml/async_work_group_copy.html):

> **This function does not perform any implicit synchronization of source data such as using a barrier before performing the copy.**

>**The kernel must wait for the completion of all async copies using the wait_group_events built-in function before exiting; otherwise the behavior is undefined.**

**Comment by L30nardoSV** [comment]
Adding a `wait_group_event` to the corresopnding last `async_work_group_copy` seem to solve the issue:

```opencl
// Copying new offspring to next generation
event_t ev = async_work_group_copy(dockpars_conformations_next + GENOTYPE_LENGTH_IN_GLOBMEM*get_group_id(0),
				   offspring_genotype,
				   dockpars_num_of_genes, 0);

// Asynchronous copy should be finished by here
wait_group_events(1,&ev);
```

**Comment by L30nardoSV** [comment]
For master, this was fixed in commit 7e56a72893449ef02bb1b341504f919f5aca845e.

**Comment by L30nardoSV** [comment]
For debugfastergrad, this was fixed in commit bd633ddd803f9f3bed575ee20e7648e43844a6ed.
